SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestAttachedComponentReplication/SpatialTestAttachedComponentReplication.cpp

```diff
@@ -0,0 +1,200 @@
+#include "SpatialTestAttachedComponentReplication.h"
+
+#include "Algo/Copy.h"
+#include "Kismet/GameplayStatics.h"
+#include "Net/UnrealNetwork.h"
+#include "SpatialFunctionalTestStep.h"
+
+ASpatialTestAttachedComponentReplicationActor::ASpatialTestAttachedComponentReplicationActor()
+{
+	bReplicates = true;
+
+	bAlwaysRelevant = true;
+}
+
+USpatialTestAttachedComponentReplicationComponent::USpatialTestAttachedComponentReplicationComponent()
+{
+	SetIsReplicatedByDefault(true);
+}
+
+void USpatialTestAttachedComponentReplicationComponent::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(ThisClass, ReplicatedValue);
+}
+
+ASpatialTestAttachedComponentReplicationActorWithDefaultComponent::ASpatialTestAttachedComponentReplicationActorWithDefaultComponent()
+{
+	CreateDefaultSubobject<USpatialTestAttachedComponentReplicationComponent>(TEXT("DefaultAttachedComponent"));
+}
+
+ASpatialTestAttachedComponentReplication::ASpatialTestAttachedComponentReplication()
+{
+	Author = TEXT("Dmitrii Kozlov <dmitriikozlov@improbable.io>");
+	Description =
+		TEXT("Check different types of attached component replication:") LINE_TERMINATOR TEXT("* Component added in a map to an actor")
+			LINE_TERMINATOR TEXT("* Component present on an actor") LINE_TERMINATOR TEXT("* Component dynamically added to an actor");
+}
+
+struct FSpatialTestAttachedComponentReplicationTypeDescription
+{
+	FString Description;
+	TSubclassOf<ASpatialTestAttachedComponentReplicationActor> TestActorClass;
+};
+
+static FSpatialTestAttachedComponentReplicationTypeDescription GetTestTypeDescription(
+	const ESpatialTestAttachedComponentReplicationType TestType)
+{
+	switch (TestType)
+	{
+	case ESpatialTestAttachedComponentReplicationType::LevelPlaced:
+		return { TEXT("Level Placed Actor"), ASpatialTestAttachedComponentReplicationActorForLevelPlacing::StaticClass() };
+	case ESpatialTestAttachedComponentReplicationType::DynamicallySpawnedWithDynamicComponent:
+		return { TEXT("Dynamic Actor with Dynamic Component"),
+				 ASpatialTestAttachedComponentReplicationActorWithDynamicComponent::StaticClass() };
+	case ESpatialTestAttachedComponentReplicationType::DynamicallySpawnedWithDefaultComponent:
+		return { TEXT("Dynamic Actor with Default Component"),
+				 ASpatialTestAttachedComponentReplicationActorWithDefaultComponent::StaticClass() };
+	default:
+		checkNoEntry();
+	}
+
+	return {};
+}
+
+static UClass* GetClassForTestType(const ESpatialTestAttachedComponentReplicationType TestType)
+{
+	return GetTestTypeDescription(TestType).TestActorClass;
+}
+
+static FString DescribeTestType(const ESpatialTestAttachedComponentReplicationType TestType)
+{
+	return GetTestTypeDescription(TestType).Description + TEXT(": ");
+}
+
+void ASpatialTestAttachedComponentReplication::PrepareTest()
+{
+	Super::PrepareTest();
+
+	GenerateTestSteps(ESpatialTestAttachedComponentReplicationType::LevelPlaced);
```

Comment:
Should I make test type a property and put three of these into the scene? They are sequentially dependent on one another right now, so should the first one fail, we won't know whether other tests work or not.

Reply:
How are they sequentially dependent? It looks like they all check against different actors currently.

Reply:
I meant "sequentially dependent" in that should one of the test cases fail (say, the LevelPlaced fails), the other two cases (dynamic component and static component) would not be tested at all.

Did fix this with the map generation thingie.